File: isegoria/src/main/java/com/eulersbridge/isegoria/util/transformation/BlurTransformation.java
package com.eulersbridge.isegoria.util.transformation;

import android.content.Context;
import android.graphics.Bitmap;
import android.support.annotation.IntRange;
import android.support.annotation.NonNull;
import android.support.v8.renderscript.Allocation;
import android.support.v8.renderscript.Element;
import android.support.v8.renderscript.RenderScript;
import android.support.v8.renderscript.ScriptIntrinsicBlur;

import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool;
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation;
import com.eulersbridge.isegoria.util.Constants;

import java.lang.ref.WeakReference;
import java.security.MessageDigest;

public class BlurTransformation extends BitmapTransformation {
    private static final String ID = String.format("%s.BlurTransformation", Constants.APP_ID);
    private static final byte[] ID_BYTES = ID.getBytes(CHARSET);

    // Blur radius in DP
    private final int blurRadius;
    public static float screenDensity = 2.0f;

    private final WeakReference<Context> weakContext;

    public BlurTransformation(Context context) {
        this(context, 25);
    }

    public BlurTransformation(Context context, @IntRange(from = 1, to = 25) int blurRadius) {
        this.weakContext = new WeakReference<>(context);
        this.blurRadius = blurRadius;
    }

    @Override
    protected Bitmap transform(@NonNull BitmapPool pool, @NonNull Bitmap toTransform, int outWidth, int outHeight) {
        Context context = weakContext.get();
        if (context == null || blurRadius <= 0) return toTransform;

        int blurRadiusPx = Math.round((float)blurRadius * screenDensity);
        if (blurRadiusPx > 25) blurRadiusPx = 25;

        if (blurRadiusPx <= 0) return toTransform;

        Bitmap outputBitmap = toTransform.copy(toTransform.getConfig(), true);

        final RenderScript renderScript = RenderScript.create(context);

        Allocation tmpIn = Allocation.createFromBitmap(renderScript, toTransform);
        Allocation tmpOut = Allocation.createFromBitmap(renderScript, outputBitmap);

        final ScriptIntrinsicBlur blur = ScriptIntrinsicBlur.create(renderScript, Element.U8_4(renderScript));
        blur.setRadius(blurRadiusPx);
        blur.setInput(tmpIn);
        blur.forEach(tmpOut);

        tmpOut.copyTo(outputBitmap);

        return outputBitmap;
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof BlurTransformation;
    }

    @Override
    public int hashCode() {
        return ID.hashCode();
    }

    @Override
    public void updateDiskCacheKey(@NonNull MessageDigest messageDigest) {
        messageDigest.update(ID_BYTES);
    }
}

File: isegoria/src/main/java/com/eulersbridge/isegoria/util/transformation/BlurTransformation.kt
package com.eulersbridge.isegoria.util.transformation

import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.support.annotation.FloatRange
import android.support.annotation.Px
import android.support.v8.renderscript.Allocation
import android.support.v8.renderscript.Element
import android.support.v8.renderscript.RenderScript
import android.support.v8.renderscript.ScriptIntrinsicBlur
import com.bumptech.glide.load.Key
import com.bumptech.glide.load.engine.bitmap_recycle.BitmapPool
import com.bumptech.glide.load.resource.bitmap.BitmapTransformation
import com.eulersbridge.isegoria.APP_ID
import java.lang.Math.min
import java.lang.Math.round
import java.lang.ref.WeakReference
import java.security.MessageDigest

class BlurTransformation @JvmOverloads constructor(
    context: Context, // Blur radius in DP
    @param:FloatRange(from = 1.0, to = 25.0) private val blurRadius: Double = 25.0
) : BitmapTransformation() {

    companion object {
        private const val ID = "$APP_ID.BlurTransformation"
        private val ID_BYTES = ID.toByteArray(Key.CHARSET)
        val screenDensity = Resources.getSystem().displayMetrics.density
    }

    private val weakContext: WeakReference<Context> = WeakReference(context)

    override fun transform(
        pool: BitmapPool,
        toTransform: Bitmap,
        outWidth: Int,
        outHeight: Int
    ): Bitmap {
        val context = weakContext.get()
        if (context == null || blurRadius <= 0)
            return toTransform

        @Px val blurRadius = min(round(blurRadius * screenDensity), 25)

        if (blurRadius <= 0)
            return toTransform

        val renderScript = RenderScript.create(context)
        val outputBitmap = toTransform.copy(toTransform.config, true)
        val tmpIn = Allocation.createFromBitmap(renderScript, toTransform)
        val tmpOut = Allocation.createFromBitmap(renderScript, outputBitmap)

        val blur = ScriptIntrinsicBlur.create(renderScript, Element.U8_4(renderScript))
        blur.apply {
            setRadius(blurRadius.toFloat())
            setInput(tmpIn)
            forEach(tmpOut)
        }

        tmpOut.copyTo(outputBitmap)

        return outputBitmap
    }

    override fun equals(other: Any?) = other is BlurTransformation

    override fun hashCode() = ID.hashCode()

    override fun updateDiskCacheKey(messageDigest: MessageDigest) = messageDigest.update(ID_BYTES)
}
